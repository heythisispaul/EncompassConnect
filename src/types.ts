/* eslint-disable camelcase */
/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */

/**
 * @packageDocumentation
 * @module Interfaces
 */
export interface SortOrderContract {
  canonicalName: string;
  order: 'asc' | 'desc';
}

export interface PipeLineRequest {
  fields?: string[];
  sortOrder?: SortOrderContract[];
}

export interface LoanGuidsPipeLineContract extends PipeLineRequest {
  loanGuids: string[];
}

export interface PipeLineTerms {
  canonicalName: string;
  matchType: 'greaterThanOrEquals' | 'exact' | 'greaterThan' | 'isNotEmpty' | 'isEmpty' | 'lessThan' | 'lessThanOrEquals' | 'equals' | 'notEquals' | 'startsWith' | 'contains';
  value?: string | number | Date;
  precision?: 'exact'| 'day' | 'month' | 'year' | 'recurring';
}

export interface PipeLineFilter {
  operator: 'and' | 'or';
  terms: (PipeLineFilter | PipeLineTerms)[];
}

export interface FilterPipeLineContract extends PipeLineRequest {
  filter: PipeLineFilter;
}

export type PipeLineContract = LoanGuidsPipeLineContract | FilterPipeLineContract;

export interface BatchLoanUpdateContract {
  filter?: PipeLineFilter;
  loanGuids?: string[];
  loanData: any;
}

export interface LoanAssociateProperties {
  loanAssociateType: 'user' | 'group';
  id: string;
  name?: string;
  phone?: string;
  cellphone?: string;
  fax?: string;
  email?: string;
  roleId?: string;
  roleName?: string;
}

export interface UserInfoContract {
  viewEmailSignature?: boolean;
  start?: number;
  limit?: number | string;
  filter?: {
    groupId?: string | number;
    roleId?: string | number;
    personaId?: string | number;
    organizationId?: string | number;
    userName?: string;
  }
}

export interface Organization {
  entityId: string | number;
  entityType: string;
  entityName: string;
  entityUri: string;
}

export interface UserProfile {
  id: string;
  lastName: string;
  firstName: string;
  fullName: string;
  email: string;
  phone: string;
  userIndicators: string[],
  peerLoanAccess: string;
  lastLogin: string | Date;
  encompassVersion: string;
  personalStatusOnline: boolean;
  personas: any[];
  workingFolder?: string;
  organization?: Organization;
  subordinateLoanAccess?: string;
  comments?: string;
  ccSite?: any;
}

export interface LicenseInformation {
  state: string;
  enabled: boolean;
  license?: string;
  expirationDate?: string | Date;
  issueDate?: string | Date;
  startDate?: string | Date;
}

export interface CreateLoanContract {
  view?: 'entity' | 'id';
  loanTemplate?: string;
  loanFolder?: string;
  loan?: any;
}

export interface EncompassConnectInitOptions {
  clientId: string;
  APIsecret: string;
  instanceId: string;
  username?: string;
  password?: string;
  version?: number;
}

export interface InternalRequestOptions {
  isRetry?: boolean;
  isNotJson?: boolean;
  version?: number;
  useTruncatedBase?: boolean;
}

export interface AssignMilestoneOptions {
  loanGuid: string;
  milestone: string;
  userId: string;
}

export interface UpdateMilestoneOptions {
  loanGuid: string;
  milestone: string;
  options: any;
  action?: 'finish' | 'unfinish';
}

export interface LoanUpdateOptions {
  appendData: boolean | string;
  persistent: 'transient' | 'permanent';
  view: 'entity' | 'id';
  loanTemplate?: string;
}

export interface BatchUpdateStatus {
  status: string;
  lastModified: string;
}

export interface BatchUpdate {
  getRequestId: () => string;
  getUpdateStatus: () => Promise<BatchUpdateStatus>;
}

export interface UpdateLoanWithGenerateContract {
  standardFields?: {
    [key: string]: any;
  },
  customFields?: {
    [key: string]: any;
  }
}

export interface TokenIntrospection {
  active: boolean;
  scope: string;
  client_id: string;
  username: string;
  token_type: 'Bearer';
  exp: number;
  sub: string;
  encompass_instance_id: string;
  user_name: string;
  user_key: string;
  encompass_user: string;
  identity_type: 'Enterprise';
  encompass_instance_type: string;
  encompass_client_id: string;
  realm_name: string;
  bearer_token?: string;
}

export interface ListOfUsersOptions {
  viewEmailSignature?: boolean;
  groupId?: string | number;
  roleId?: string;
  personaId?: string | number;
  organizationId?: string | number;
  userName?: string;
  start?: number;
  limit?: number;
}
