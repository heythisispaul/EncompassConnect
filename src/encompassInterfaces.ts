/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */
export interface sortOrderContract {
  canonicalName: string;
  order: 'asc' | 'desc';
}

export interface PipeLineRequest {
  fields?: string[];
  sortOrder?: sortOrderContract[];
}

export interface LoanGuidsPipeLineContract extends PipeLineRequest {
  loanGuids: string[];
}

export interface PipeLineTerms {
  canonicalName: string;
  matchType: 'greaterThanOrEquals' | 'exact' | 'greaterThan' | 'isNotEmpty' | 'isEmpty' | 'lessThan' | 'lessThanOrEquals' | 'equals' | 'notEquals' | 'startsWith' | 'contains';
  value?: string | number | Date;
  precision?: 'exact'| 'day' | 'month' | 'year' | 'recurring';
}

export interface PipeLineFilter {
  operator: 'and' | 'or';
  terms: PipeLineTerms[];
}

export interface FilterPipeLineContract extends PipeLineRequest {
  filter: PipeLineFilter;
}

export type PipeLineContract = LoanGuidsPipeLineContract | FilterPipeLineContract;

export interface BatchLoanUpdateContract {
  filter?: PipeLineFilter;
  loanGuids?: string[];
  loanData: any;
}

export interface LoanAssociateProperties {
  loanAssociateType: 'user' | 'group';
  id: string;
  name?: string;
  phone?: string;
  cellphone?: string;
  fax?: string;
  email?: string;
  roleId?: string;
  roleName?: string;
}

export interface UserInfoContract {
  viewEmailSignature?: boolean;
  start?: number;
  limit?: number | string;
  filter?: {
    groupId?: string | number;
    roleId?: string | number;
    personaId?: string | number;
    organizationId?: string | number;
    userName?: string;
  }
}

export interface Organization {
  entityId: string | number;
  entityType: string;
  entityName: string;
  entityUri: string;
}

export interface UserProfile {
  id: string;
  lastName: string;
  firstName: string;
  fullName: string;
  email: string;
  phone: string;
  userIndicators: string[],
  peerLoanAccess: string;
  lastLogin: string | Date;
  encompassVersion: string;
  personalStatusOnline: boolean;
  personas: any[];
  workingFolder?: string;
  organization?: Organization;
  subordinateLoanAccess?: string;
  comments?: string;
  ccSite?: any;
}

export interface LicenseInformation {
  state: string;
  enabled: boolean;
  license?: string;
  expirationDate?: string | Date;
  issueDate?: string | Date;
  startDate?: string | Date;
}

export interface CreateLoanContract {
  view?: 'entity' | 'id';
  loanTemplate?: string;
  loanFolder?: string;
  loan?: any;
}

export interface EncompassConnectConstructor {
  clientId: string;
  APIsecret: string;
  instanceId: string;
  username?: string;
  password?: string;
}

export interface InternalRequestOptions {
  isRetry?: boolean;
  isNotJson?: boolean;
}

export interface AssignMilestone {
  loanGuid: string;
  milestone: string;
  userId: string;
}

export interface UpdateMilestone {
  loanGuid: string;
  milestone: string;
  options: any;
  action?: 'finish' | 'unfinish';
}

export interface LoanUpdateOptions {
  appendData: boolean;
  persistent: 'transient' | 'permanent';
  view: 'entity' | 'id';
  loanTemplate?: string;
}
