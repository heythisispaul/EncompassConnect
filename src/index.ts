import EncompassConnect from './encompassConnect';

export type { Response, RequestInit } from 'node-fetch';
export type {
  AssignMilestoneOptions,
  BatchLoanUpdateContract,
  EncompassConnectInitOptions,
  PipeLineContract,
  UpdateMilestoneOptions,
  LoanUpdateOptions,
  BatchUpdateStatus,
  UpdateLoanWithGenerateContract,
  BatchUpdate,
  TokenIntrospection,
  ListOfUsersOptions,
} from './types';

export default EncompassConnect;
