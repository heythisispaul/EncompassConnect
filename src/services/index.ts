export { default as LoanService } from './loans';
export { default as MilestoneService } from './milestones';
export { default as SchemaService } from './schemas';
export { default as UserService } from './users';
