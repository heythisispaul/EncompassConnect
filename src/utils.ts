export const massageCustomFields = (customFields: any) => {
  if (!customFields) {
    return [];
  }
  const massaged = Object
    .entries(customFields).map(([fieldName, value]: [string, any]) => {
      const parsedNumber = parseFloat(value);
      const customFieldObject: any = {
        fieldName,
        stringValue: value && value.toString ? value.toString() : value,
      };
      if (!Number.isNaN(parsedNumber)) {
        customFieldObject.numericValue = parsedNumber;
      }
      return customFieldObject;
    });
  return massaged;
};

export const objectToURLString = (input: any, queryString?: boolean): string => {
  // @ts-ignore
  const result = input ? new URLSearchParams(input).toString() : '';
  return queryString ? `?${result}` : result;
};
