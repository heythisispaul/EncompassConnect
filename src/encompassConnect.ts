/* eslint-disable max-len */
/* eslint-disable camelcase */

/**
 * @packageDocumentation
 * @module EncompassConnectClass
 */
import fetch, {
  Response,
  RequestInit,
  HeadersInit,
} from 'node-fetch';
import {
  EncompassConnectInitOptions,
  InternalRequestOptions,
  PipeLineContract,
  BatchLoanUpdateContract,
  BatchUpdateStatus,
  BatchUpdate,
  TokenIntrospection,
} from './types';
import {
  LoanService,
  MilestoneService,
  SchemaService,
  UserService,
} from './services';
import { objectToURLString } from './utils';

class EncompassConnect {
  /**
   * The client ID of the Encompass instance to connect to.
   */
  #clientId: string;

  /**
   * The API key of the Encompass instance to connect to.
   */
  #APIsecret: string;

  /**
   * The Encompass instance ID of the instance to connect to.
   */
  #instanceId: string;

  /**
   * The password of the account to retrieve tokens with. Only stores the value provided to the constructor.
   */
  #password: string;

  /**
   * The bearer token that is used to authenticate each request to the Encompas API. The same token will be reused
   * until either a `401` is returned, or the value is overwritten with the `setToken()` or `getToken()` methods.
   */
  #token: string | null;

  /**
   * The username of the account to retrieve tokens with. Only stores the value provided to the constructor.
   */
  username: string;

  /**
   * The domain of the Encompass API.
   */
  base: string;

  loans: LoanService;

  milestones: MilestoneService;

  schemas: SchemaService;

  users: UserService;

  /**
   * @ignore
   */
  authBase: string;

  /**
   * The version of the Encompass API to send requests to. Defaults to `1`.
   */
  version: number;

  constructor({
    clientId,
    APIsecret,
    instanceId,
    username,
    password,
    version = 1,
  }: EncompassConnectInitOptions) {
    this.#clientId = clientId;
    this.#APIsecret = APIsecret;
    this.#instanceId = instanceId;
    this.#password = password || '';
    this.#token = '';
    this.username = username || '';
    this.version = version;
    this.base = 'https://api.elliemae.com/encompass';
    this.authBase = 'https://api.elliemae.com';
    this.loans = new LoanService(this);
    this.milestones = new MilestoneService(this);
    this.schemas = new SchemaService(this);
    this.users = new UserService(this);
  }

  /**
   * Replaces the `#token` property with the provided token value. The instance can be implicitly 'logged out' by setting this value to `null` (if it was not provided a username and password in the constructor).
   */
  setToken(token:string | null): void {
    this.#token = token;
  }

  /**
   * @ignore
   */
  private withTokenHeader(headers = {}): HeadersInit {
    return {
      ...headers,
      Authorization: `Bearer ${this.#token}`,
    };
  }

  /**
   * @ignore
   */
  async fetchWithRetry(
    path: string,
    options: RequestInit = {},
    customOptions: InternalRequestOptions = {},
  ): Promise<any> {
    const { isRetry, isNotJson, version } = customOptions;
    const shouldRetry = !isRetry && this.username && this.#password;
    const failedAuthError = new Error(
      `Token invalid. ${!isRetry && shouldRetry ? 'Will reattempt with new token' : 'Unable to get updated one.'}`,
    );
    try {
      if (!this.#token) {
        await this.getToken();
      }
      const url = `${this.base}/v${version || this.version}${path}`;
      const optionsWithToken: RequestInit = {
        ...options,
        headers: this.withTokenHeader(options.headers),
      };
      const response: Response = await fetch(url, optionsWithToken);
      if (response.status === 401) {
        throw failedAuthError;
      }
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return isNotJson ? response : await response.json();
    } catch (error) {
      if (shouldRetry && error === failedAuthError) {
        this.setToken(null);
        return this.fetchWithRetry(path, options, {
          ...customOptions,
          isRetry: true,
        });
      }
      throw error;
    }
  }

  /**
   * Exchanges the provided username and password for a bearer token and stores it to the `#token` property of the instance.
   * If no username and password are provided, it will fallback to the username and password values provided to the constructor.
   */
  async getToken(username?: string, password?: string): Promise<void> {
    const body: string = objectToURLString({
      grant_type: 'password',
      username: `${username || this.username}@encompass:${this.#instanceId}`,
      password: password || this.#password,
      client_id: this.#clientId,
      client_secret: this.#APIsecret,
    });

    const requestOptions: RequestInit = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body,
      redirect: 'follow',
    };

    const response = await fetch(`${this.authBase}/oauth2/v1/token`, requestOptions);
    const { access_token } = await response.json();
    this.setToken(access_token);
  }

  /**
   * Calls the token introspection API with the provided token. If a token is not provided, it will introspect the token stored to the `#token` property of the instance as a fallback.
   * If the introspection returns a valid token, it will return the response body of the request, if the token is invalid, returns `null`.
   */
  async introspectToken(token?: string): Promise<TokenIntrospection | null> {
    const body: string = objectToURLString({
      token: token || this.#token,
    });
    const Authorization: string = `Basic ${Buffer.from(`${this.#clientId}:${this.#APIsecret}`).toString('base64')}`;
    const requestOptions: RequestInit = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        Authorization,
      },
      body,
    };
    try {
      const response = await fetch(`${this.authBase}/oauth2/v1/token/introspection`, requestOptions);
      if (response.ok) {
        const tokenData: TokenIntrospection = await response.json();
        return tokenData;
      }
      return null;
    } catch (error) {
      return null;
    }
  }

  /**
   * Returns the result of the 'Get Canonical Names' endpoint.
   */
  async getCanonicalNames(): Promise<any> {
    const canonicalNames = await this.fetchWithRetry('/loanPipeline/fieldDefinitions');
    return canonicalNames;
  }

  /**
    *  Generate a pipeline view by calling the `viewPipeline()` method. This method has one required argument, a `PipeLineContract`, and can optionally take a limit value as the second argument:
    *
    *  ```typescript
    *  // a pipelineContract expects either a loanGuids array, or a filter object:
    *  const commonFilterValues = {
    *    sortOrder: [
    *      {
    *        canonicalName: 'Loan.LastModified',
    *        order: 'desc'
    *      }
    *    ],
    *    fields: [
    *      "Loan.LoanAmount",
    *      "Fields.4002"
    *    ],
    *  };
    *
    *  const pipelineWithGuids: LoanGuidsPipeLineContract = {
    *    ...commonFilterValues,
    *    loanGuids: [
    *      'some-loan-guid-1',
    *      'some-loan-guid-2',
    *    ],
    *  };
    *
    *  const pipelineWithFilter: FilterPipeLineContract = {
    *    ...commonFilterValues,
    *    filter: {
    *      operator: 'and',
    *      terms: [
    *        {
    *          canonicalName: "Loan.LastModified",
    *          matchType: "greaterThanOrEquals",
    *          value: new Date()
    *        },
    *        {
    *          canonicalName: "Loan.LoanFolder",
    *          matchType: "exact",
    *          value: "My Pipeline"
    *        }
    *      ]
    *    },
    *  };
    *
    *  const pipelineDataFromGuids = await encompass.viewPipeline(pipelineWithGuids);
    *
    *  // or with the other contract, and a limit of the first 50 results:
    *  const pipelineDataFromFilter = await encompass.viewPipeline(pipelineWithFilter, 50);
    *  ```
    */
  async viewPipeline(options: PipeLineContract, limit?: number) {
    const uri = `/loanPipeline${limit ? `?limit=${limit}` : ''}`;
    const pipeLineData = await this.fetchWithRetry(uri, {
      method: 'POST',
      headers: this.withTokenHeader(),
      body: JSON.stringify(options),
    });
    return pipeLineData;
  }

  /**
    *  The batch update API allows your to apply the same loan data to multiple loans and can be invoked with `batchLoanUpdate()` method.
    *
    *  This method returns an object that with it's own functionality to check the status of batch update, or to get the request ID if needed.
    *
    *  Just like viewing a pipeline, either a filter or an array of lan GUIDs can be provided.
    *  ```typescript
    *  const updateData: BatchLoanUpdateContract = {
    *    loanGuids: [
    *      // array of loan GUIDs to update
    *    ],
    *    loanData: {
    *      // contract of the loan data to apply to each loan
    *    },
    *  };
    *
    *  const exampleBatchUpdate: BatchUpdate = await encompass.batchLoanUpdate(updateData);
    *
    *  // the return value can be used to check the status:
    *  const latestStatus: BatchUpdateStatus = await exampleBatchUpdate.getUpdateStatus();
    *  console.log(latestStatus.status) // 'done' or 'error'
    *
    *  // or if needed you can get the request ID itself:
    *  const exampleBatchUpdateId: string = exampleBatchUpdate.getRequestId();
    *  ```
    */
  async batchLoanUpdate(options: BatchLoanUpdateContract): Promise<BatchUpdate> {
    const response: any = await this.fetchWithRetry('/loanBatch/updateRequests', {
      method: 'POST',
      body: JSON.stringify(options),
    }, { isNotJson: true });
    const requestId: string = response && response.headers
      ? response.headers.get('location').split('/').reverse()[0]
      : null;
    return {
      getRequestId: () => requestId,
      getUpdateStatus: async () => {
        const url = `/loanBatch/updateRequests/${requestId}`;
        const status: BatchUpdateStatus = await this.fetchWithRetry(url);
        return status;
      },
    };
  }

  /**
    * If an API is not available through an explicit method in this class, the `request()` method is a wrapper around fetch that allows you to request any Encompass API endpoint.
    *
    * It takes in the same arguments as any [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) API, with the exception that the first argument is appended as a path to the Encompass API domain.
    * ```typescript
    * // hitting the Get Custom Fields API:
    * const customFieldsResponse: Response = await encompass.request('/settings/loan/customFields');
    * const data = await customFieldsResponse.json();
    * console.log(data);
    *
    * // or update a contact:
    * const options: RequestInit = {
    *   method: 'POST',
    *   body: {
    *     firstname: 'contact first name',
    *     lastname: 'contact last name',
    *   },
    * };
    * await encompass.request('/businessContacts/<some-contact-id>', options);
    * ```
    */
  async request(url: string, options: RequestInit): Promise<Response> {
    const response = await this.fetchWithRetry(url, options, { isNotJson: true });
    return response;
  }
}

export default EncompassConnect;
