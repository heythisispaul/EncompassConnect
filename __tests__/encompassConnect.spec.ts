import EncompassConnect from '../src/encompassConnect';
import { BatchLoanUpdateContract } from '../src/types';
import {
  mockFetchError,
  mockResponse,
  mockResponseTimes,
  mockResponseJson,
  fetch,
} from './__utils__/mockFunctions';
import {
  testInstanceWithToken,
  testInstanceWithCreds,
  testInstanceWithOnAuth,
  testInstance,
  mockToken,
  defaultCreds,
  createConstructor,
  mockOnAuthenticate,
} from './__utils__/mockEncompassConnectInstances';
import {
  returnsResponseBody,
  matchesFetchSnapshot,
} from './__utils__/commonTests';

jest.mock('node-fetch');

describe('EncompassConnect', () => {
  beforeEach(() => {
    fetch.mockReset();
    mockResponseJson.mockReset();
    mockOnAuthenticate.mockReset();
  });

  describe('getTokenWithCredentials', () => {
    it('uses the username and password from the constructor to create a token if none are provided', async () => {
      mockResponse({ access_token: mockToken });
      await testInstanceWithCreds.getTokenWithCredentials();
      expect(fetch.mock.calls[0]).toMatchSnapshot();
    });

    it('uses the provided username and password over the ones in the constructor to create a token', async () => {
      mockResponse({ access_token: mockToken });
      await testInstance.getTokenWithCredentials('PROVIDED_USERNAME', 'PROVIDED_PASSWORD');
      expect(fetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('introspectToken', () => {
    it('returns the body of the response if the call was successful', async () => {
      const mockIntrospectResponse = 'mock introspect response';
      mockResponse(mockIntrospectResponse);
      const result = await testInstanceWithCreds.introspectToken();
      expect(result).toEqual(mockIntrospectResponse);
    });

    it('returns null if the token is invalid', async () => {
      mockResponse({ error: 'token invalid' }, 400);
      const result = await testInstanceWithCreds.introspectToken();
      expect(result).toEqual(null);
    });

    it('returns null if an exception occurs', async () => {
      mockFetchError();
      const result = await testInstanceWithCreds.introspectToken();
      expect(result).toEqual(null);
    });

    it('adds an Authorization header to the fetch request', async () => {
      await testInstanceWithCreds.introspectToken();
      // @ts-ignore
      const { headers } = fetch.mock.calls[0][1];
      // @ts-ignore
      expect(<string>headers.Authorization).toBeTruthy();
    });

    it('introspects the token provided over the stored token', async () => {
      const testToken = 'TEST_TOKEN_VALUE';
      await testInstanceWithCreds.introspectToken(testToken);
      // @ts-ignore
      const { body } = fetch.mock.calls[0][1];
      expect(body).toEqual(`token=${testToken}`);
    });
  });

  describe('getToken', () => {
    it('returns the #token value of the instance', () => {
      const token = testInstanceWithToken.getToken();
      expect(token).toEqual(mockToken);
    });
  });

  describe('getCanonicalNames', () => {
    it('requests the correct endpoint', async () => {
      mockResponse();
      await testInstanceWithToken.getCanonicalNames();
      // @ts-ignore
      const path: string = fetch.mock.calls[0][0];
      expect(path.includes('/loanPipeline/fieldDefinitions')).toBeTruthy();
    });

    it('fetches with the correct request options', async () => (
      matchesFetchSnapshot('getCanonicalNames')
    ));

    it('returns the body of the response', async () => (
      returnsResponseBody('getCanonicalNames')
    ));
  });

  describe('viewPipeline', () => {
    const defaultOptions = {
      filter: {
        operator: 'and' as any,
        terms: [],
      },
      fields: ['test.field'],
    };
    it('makes the API call with the correct information', async () => (
      matchesFetchSnapshot('viewPipeline', defaultOptions)
    ));

    it('adds the limit query string if provided', async () => {
      mockResponse();
      await testInstanceWithToken.viewPipeline(defaultOptions, 100);
      // @ts-ignore
      const path: string = fetch.mock.calls[0][0];
      expect(path.includes('?limit=100')).toBeTruthy();
    });

    it('returns the body of the response', async () => (
      returnsResponseBody('viewPipeline', defaultOptions)
    ));
  });

  describe('batchLoanUpdate', () => {
    const defaultOptions: BatchLoanUpdateContract = {
      loanGuids: ['test-1', 'test-2'],
      loanData: {
        key1: 'value1',
        key2: 'value2',
      },
    };

    it('makes the API call with the correct information', async () => (
      matchesFetchSnapshot('batchLoanUpdate', defaultOptions)
    ));

    it('returns a BatchUpdate object with the request id', async () => {
      const id = '12345';
      mockResponse({}, 200, { location: `some/path/${id}` });
      const batchUpdate = await testInstanceWithToken.batchLoanUpdate(defaultOptions);
      expect(batchUpdate.getRequestId()).toEqual(id);
    });

    it('returns a BatchUpdate object that can fetch its status', async () => {
      const id = '9876';
      mockResponse({}, 200, { location: `some/path/${id}` });
      mockResponse();
      const batchUpdate = await testInstanceWithToken.batchLoanUpdate(defaultOptions);
      await batchUpdate.getUpdateStatus();
      expect(fetch.mock.calls[1]).toMatchSnapshot();
    });
  });

  describe('request', () => {
    it('fetches with the correct information', async () => {
      mockResponse();
      await testInstanceWithToken.request('test/url', { method: 'POST', body: JSON.stringify({ some: 'value' }) });
      expect(fetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('fetchWithRetry', () => {
    const createNewInstanceWithToken = (excludeCreds?: boolean, tokenValue: string = 'some token value'): EncompassConnect => {
      const newInstanceWithToken = new EncompassConnect(
        createConstructor(excludeCreds ? {} : defaultCreds),
      );
      newInstanceWithToken.setToken(tokenValue);
      return newInstanceWithToken;
    };

    it('fetches a token if one is not already set in the instance', async () => {
      expect.assertions(2);
      mockResponseTimes(2);
      const newInstance = new EncompassConnect(createConstructor(defaultCreds));
      await newInstance.getCanonicalNames();
      expect(fetch).toHaveBeenCalledTimes(2);
      const firstRequestedUrl = fetch.mock.calls[0][0];
      expect(firstRequestedUrl).toEqual(`${testInstanceWithCreds.authBase}/oauth2/v1/token`);
    });

    it('gets a fetches a token if the response was a 401 status and it has credentials', async () => {
      mockResponse({}, 401);
      mockResponseTimes(2);
      const newInstanceWithBadToken = createNewInstanceWithToken();
      await newInstanceWithBadToken.getCanonicalNames();
      expect(fetch).toHaveBeenCalledTimes(3);
      const firstRequestedUrl = fetch.mock.calls[1][0];
      expect(firstRequestedUrl).toEqual(`${testInstanceWithCreds.authBase}/oauth2/v1/token`);
    });

    it('retries the same request if it successfully gets a new token', async () => {
      mockResponse({}, 401);
      mockResponseTimes(2);
      const newInstanceWithBadToken = createNewInstanceWithToken();
      await newInstanceWithBadToken.getCanonicalNames();
      const [firstCall, retriedCall] = [0, 2].map((number) => fetch.mock.calls[number][0]);
      expect(firstCall).toEqual(retriedCall);
    });

    it('throws an error if it receives a non-401 but invalid status code', async () => {
      mockResponse({}, 500);
      const instanceWithToken = createNewInstanceWithToken();
      await expect(instanceWithToken.getCanonicalNames())
        .rejects.toThrowError();
      expect(fetch).toHaveBeenCalledTimes(1);
    });

    it('does not retry if it receives a 401 but does not have a credentials', async () => {
      mockResponseTimes(3, {}, 401);
      const instanceWithToken = createNewInstanceWithToken(true);
      await expect(instanceWithToken.getCanonicalNames())
        .rejects.toThrowError();
      expect(fetch).toHaveBeenCalledTimes(1);
    });

    it('deletes the invalid token if it receives a 401', async () => {
      mockResponseTimes(3, {}, 401);
      const instanceWithToken = createNewInstanceWithToken();
      await expect(instanceWithToken.getCanonicalNames())
        .rejects.toThrowError();
      const firstCallAfterFailure = fetch.mock.calls[1][0];
      expect(firstCallAfterFailure).toEqual(`${testInstanceWithCreds.authBase}/oauth2/v1/token`);
    });

    it('does not attempt to parse the body if provided in the config', async () => {
      mockResponseTimes(3);
      const instanceWithToken = createNewInstanceWithToken();
      await instanceWithToken.request('someurl');
      expect(mockResponseJson).not.toHaveBeenCalled();
    });

    it('calls the onAuthenticated function if provided', async () => {
      mockResponseTimes(3);
      await testInstanceWithOnAuth.request('someurl');
      expect(mockOnAuthenticate).toHaveBeenCalled();
    });
  });
});
