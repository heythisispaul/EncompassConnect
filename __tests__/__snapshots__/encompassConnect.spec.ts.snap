// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EncompassConnect batchLoanUpdate makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loanBatch/updateRequests",
  Object {
    "body": "{\\"loanGuids\\":[\\"test-1\\",\\"test-2\\"],\\"loanData\\":{\\"key1\\":\\"value1\\",\\"key2\\":\\"value2\\"}}",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "POST",
  },
]
`;

exports[`EncompassConnect batchLoanUpdate returns a BatchUpdate object that can fetch its status 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loanBatch/updateRequests/9876",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;

exports[`EncompassConnect getCanonicalNames fetches with the correct request options 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loanPipeline/fieldDefinitions",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;

exports[`EncompassConnect getToken uses the provided username and password over the ones in the constructor to create a token 1`] = `
Array [
  "https://api.elliemae.com/oauth2/v1/token",
  Object {
    "body": "grant_type=password&username=PROVIDED_USERNAME%40encompass%3A%3CINSTANCE+ID%3E&password=PROVIDED_PASSWORD&client_id=%3CCLIENT+ID%3E&client_secret=%3CAPI+SECRET%3E",
    "headers": Object {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
    "redirect": "follow",
  },
]
`;

exports[`EncompassConnect getToken uses the username and password from the constructor to create a token if none are provided 1`] = `
Array [
  "https://api.elliemae.com/oauth2/v1/token",
  Object {
    "body": "grant_type=password&username=%3CCONSTRUCTOR+USERNAME%3E%40encompass%3A%3CINSTANCE+ID%3E&password=%3CCONSTRUCTOR+PASSWORD%3E&client_id=%3CCLIENT+ID%3E&client_secret=%3CAPI+SECRET%3E",
    "headers": Object {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
    "redirect": "follow",
  },
]
`;

exports[`EncompassConnect request fetches with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1test/url",
  Object {
    "body": "{\\"some\\":\\"value\\"}",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "POST",
  },
]
`;

exports[`EncompassConnect viewPipeline makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loanPipeline",
  Object {
    "body": "{\\"filter\\":{\\"operator\\":\\"and\\",\\"terms\\":[]},\\"fields\\":[\\"test.field\\"]}",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "POST",
  },
]
`;
